File: src/pgc/InputData.java
package pgc;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author Thomas Povinelli
 * Created 2019-Jan-16
 * In Pressure-Gradient-Calculator-for-MS-Stars
 */
public class InputData {
  static class Parameters {
    private int limit;
    private long radius, steps;
    private double density;

    public Parameters(final int limit, final long radius, final long steps, final double density) {
      this.limit = limit;
      this.radius = radius;
      this.steps = steps;
      this.density = density;
    }

    public int getLimit() {
      return limit;
    }

    public long getRadius() {
      return radius;
    }

    public long getSteps() {
      return steps;
    }

    public double getDensity() {
      return density;
    }
  }

  private String limitText;
  private String radiusText;
  private String densityText;
  private String stepText;

  public InputData(
      @NotNull final String limitText,
      @NotNull final String radiusText,
      @NotNull final String densityText,
      @NotNull final String stepText
  ) {
    this.limitText = limitText.trim();
    this.radiusText = radiusText.trim();
    this.densityText = densityText.trim();
    this.stepText = stepText.trim();
  }

  @Nullable
  public Parameters parse() {
    final int limit;
    if (limitText.equals("")) {
      limit = 45;
    } else {
      try {
        limit = Integer.parseInt(limitText);
      } catch (NumberFormatException e1) {
        Controller.displayAlert("Please enter a valid integer for limit");
        return null;
      }
    }
    final long radius;
    if (radiusText.equals("")) {
      radius = 23193333L;
    } else {
      try {
        radius = Long.parseLong(radiusText);
      } catch (NumberFormatException e2) {
        Controller.displayAlert("Please enter a valid long integer for radius");
        return null;
      }
    }
    final long steps;
    if (stepText.equals("")) {
      steps = 45000000L;
    } else {
      try {
        steps = Long.parseLong(stepText);
      } catch (NumberFormatException e1) {
        Controller.displayAlert("Please enter a valid long integer for steps");
        return null;
      }
    }
    final double density;
    if (densityText.equals("")) {
      density = 4.5;
    } else {
      try {
        density = Double.parseDouble(densityText);
      } catch (NumberFormatException e1) {
        Controller.displayAlert("Please enter a valid double precision floating point number for density");
        return null;
      }
    }
    return new Parameters(limit, radius, steps, density);
  }

}

File: src/pgc/InputData.kt
package pgc

import java.lang.reflect.InvocationTargetException

/**
 * @author Thomas Povinelli
 * Created 2019-Jan-16
 * In Pressure-Gradient-Calculator-for-MS-Stars
 */
class InputData(
  limitText: String, radiusText: String,
  densityText: String, stepText: String
) {

  private val limitText: String = limitText.trim { it <= ' ' }
  private val radiusText: String = radiusText.trim { it <= ' ' }
  private val densityText: String = densityText.trim { it <= ' ' }
  private val stepText: String = stepText.trim { it <= ' ' }

  data class Parameters(
    val limit: Int, val radius: Long,
    val steps: Long, val density: Double
  )

  fun parse(): Parameters? {
    val limit = parseInputText(limitText, 45, "sections") ?: return null
    val radius = parseInputText(radiusText, 23193333L, "radius") ?: return null
    val steps = parseInputText(stepText, 45000000L, "steps") ?: return null
    val density = parseInputText(densityText, 4.5, "density") ?: return null
    return Parameters(limit, radius, steps, density)
  }

  private inline fun <reified T : Number> parseInputText(
    inputText: String,
    defaultValue: T,
    fieldName: String
  ): T? {
    return if (inputText == "") {
      defaultValue
    } else {
      try {
        val p = T::class.java.getDeclaredMethod("parse${T::class.java.simpleName.fixMe()}", String::class.java)
        p.invoke(null, inputText) as T
      } catch (e: InvocationTargetException) {
        Controller.displayAlert("Please enter a valid " + T::class.java.simpleName + " for " + fieldName)
        return null
      }
    }
  }

  // WHY IS parseInt THE *ONLY* METHOD THAT DOES NOT END IN ITS CLASS NAME?!
  private fun String.fixMe(): String = if (this == "Integer") "Int" else this
}


